File: packages/schema/src/typescript/crm.ts
/**
 * CRM schema definitions for Transform Army AI Adapter Service.
 * 
 * This module defines Zod schemas for CRM operations including contacts,
 * companies, deals, and notes across different CRM providers.
 */

import { z } from 'zod';
import { ToolInputSchema, PaginationParamsSchema, PaginationResponseSchema } from './base';

/**
 * CRM contact model representing a person in the CRM system.
 */
export const ContactSchema = z.object({
  id: z.string().describe('Unique contact identifier'),
  email: z.string().email().describe('Contact email address'),
  first_name: z.string().optional().describe('Contact first name'),
  last_name: z.string().optional().describe('Contact last name'),
  company: z.string().optional().describe('Company name'),
  phone: z.string().optional().describe('Phone number'),
  title: z.string().optional().describe('Job title'),
  url: z.string().url().optional().describe('URL to view contact in provider system'),
  created_at: z.string().datetime().or(z.date()).optional().describe('Contact creation timestamp'),
  updated_at: z.string().datetime().or(z.date()).optional().describe('Last update timestamp'),
  custom_fields: z.record(z.any()).optional().describe('Provider-specific custom fields'),
  owner_id: z.string().optional().describe('ID of the contact owner/assignee'),
  tags: z.array(z.string()).optional().describe('Tags associated with the contact'),
});
export type Contact = z.infer<typeof ContactSchema>;

/**
 * CRM company/account model.
 */
export const CompanySchema = z.object({
  id: z.string().describe('Unique company identifier'),
  name: z.string().describe('Company name'),
  domain: z.string().optional().describe('Company website domain'),
  industry: z.string().optional().describe('Industry/sector'),
  employees: z.number().int().min(0).optional().describe('Number of employees'),
  annual_revenue: z.number().min(0).optional().describe('Annual revenue'),
  phone: z.string().optional().describe('Company phone number'),
  address: z.string().optional().describe('Company address'),
  city: z.string().optional().describe('City'),
  state: z.string().optional().describe('State/province'),
  country: z.string().optional().describe('Country'),
  postal_code: z.string().optional().describe('Postal/ZIP code'),
  url: z.string().url().optional().describe('URL to view company in provider system'),
  created_at: z.string().datetime().or(z.date()).optional().describe('Company creation timestamp'),
  updated_at: z.string().datetime().or(z.date()).optional().describe('Last update timestamp'),
  custom_fields: z.record(z.any()).optional().describe('Provider-specific custom fields'),
});
export type Company = z.infer<typeof CompanySchema>;

/**
 * CRM deal/opportunity model.
 */
export const DealSchema = z.object({
  id: z.string().describe('Unique deal identifier'),
  name: z.string().describe('Deal name'),
  amount: z.number().min(0).optional().describe('Deal amount/value'),
  currency: z.string().length(3).default('USD').describe('Currency code (ISO 4217)'),
  stage: z.string().describe('Current deal stage/status'),
  probability: z.number().min(0).max(1).optional().describe('Win probability (0-1)'),
  close_date: z.string().optional().describe('Expected close date (YYYY-MM-DD)'),
  contact_ids: z.array(z.string()).optional().describe('Associated contact IDs'),
  company_id: z.string().optional().describe('Associated company ID'),
  owner_id: z.string().optional().describe('Deal owner/assignee ID'),
  url: z.string().url().optional().describe('URL to view deal in provider system'),
  created_at: z.string().datetime().or(z.date()).optional().describe('Deal creation timestamp'),
  updated_at: z.string().datetime().or(z.date()).optional().describe('Last update timestamp'),
  custom_fields: z.record(z.any()).optional().describe('Provider-specific custom fields'),
});
export type Deal = z.infer<typeof DealSchema>;

/**
 * CRM note/activity model.
 */
export const NoteSchema = z.object({
  id: z.string().describe('Unique note identifier'),
  content: z.string().describe('Note content/body'),
  type: z.string().optional().describe("Note type (e.g., 'call_note', 'email', 'meeting')"),
  contact_id: z.string().optional().describe('Associated contact ID'),
  company_id: z.string().optional().describe('Associated company ID'),
  deal_id: z.string().optional().describe('Associated deal ID'),
  author_id: z.string().optional().describe('Note author/creator ID'),
  timestamp: z.string().datetime().or(z.date()).optional().describe('Note timestamp (when the activity occurred)'),
  created_at: z.string().datetime().or(z.date()).optional().describe('Note creation timestamp in system'),
});
export type Note = z.infer<typeof NoteSchema>;

/**
 * Contact data for creation.
 */
export const ContactDataSchema = z.object({
  email: z.string().email().describe('Contact email (required)'),
  first_name: z.string().optional().describe('First name'),
  last_name: z.string().optional().describe('Last name'),
  company: z.string().optional().describe('Company name'),
  phone: z.string().optional().describe('Phone number'),
  title: z.string().optional().describe('Job title'),
  owner_id: z.string().optional().describe('Owner ID'),
  tags: z.array(z.string()).optional().describe('Tags'),
  custom_fields: z.record(z.any()).optional().describe('Custom fields'),
});
export type ContactData = z.infer<typeof ContactDataSchema>;

/**
 * Options for contact creation.
 */
export const ContactOptionsSchema = z.object({
  dedupe_by: z.array(z.string()).optional().describe("Fields to check for duplicates (e.g., ['email'])"),
  update_if_exists: z.boolean().default(false).describe('Update contact if duplicate found'),
});
export type ContactOptions = z.infer<typeof ContactOptionsSchema>;

/**
 * Request to create a new CRM contact.
 */
export const CreateContactRequestSchema = ToolInputSchema.extend({
  contact: ContactDataSchema.describe('Contact data to create'),
  options: ContactOptionsSchema.optional().describe('Creation options'),
});
export type CreateContactRequest = z.infer<typeof CreateContactRequestSchema>;

/**
 * Request to update an existing CRM contact.
 */
export const UpdateContactRequestSchema = ToolInputSchema.extend({
  updates: z.record(z.any()).describe('Fields to update (partial update)'),
});
export type UpdateContactRequest = z.infer<typeof UpdateContactRequestSchema>;

/**
 * Deal data for creation.
 */
export const DealDataSchema = z.object({
  name: z.string().describe('Deal name'),
  amount: z.number().min(0).optional().describe('Deal amount'),
  currency: z.string().length(3).default('USD').describe('Currency code'),
  stage: z.string().describe('Deal stage'),
  probability: z.number().min(0).max(1).optional().describe('Win probability'),
  close_date: z.string().optional().describe('Expected close date (YYYY-MM-DD)'),
  contact_ids: z.array(z.string()).optional().describe('Associated contact IDs'),
  company_id: z.string().optional().describe('Associated company ID'),
  owner_id: z.string().optional().describe('Deal owner ID'),
  custom_fields: z.record(z.any()).optional().describe('Custom fields'),
});
export type DealData = z.infer<typeof DealDataSchema>;

/**
 * Request to create a new CRM deal/opportunity.
 */
export const CreateDealRequestSchema = ToolInputSchema.extend({
  deal: DealDataSchema.describe('Deal data to create'),
});
export type CreateDealRequest = z.infer<typeof CreateDealRequestSchema>;

/**
 * Request to update an existing CRM deal.
 */
export const UpdateDealRequestSchema = ToolInputSchema.extend({
  updates: z.record(z.any()).describe('Fields to update (partial update)'),
});
export type UpdateDealRequest = z.infer<typeof UpdateDealRequestSchema>;

/**
 * Note data for creation.
 */
export const NoteDataSchema = z.object({
  content: z.string().describe('Note content'),
  type: z.string().optional().describe('Note type'),
  timestamp: z.string().datetime().or(z.date()).optional().describe('When the activity occurred'),
});
export type NoteData = z.infer<typeof NoteDataSchema>;

/**
 * Request to add a note to a CRM entity.
 */
export const AddNoteRequestSchema = ToolInputSchema.extend({
  note: NoteDataSchema.describe('Note data to create'),
});
export type AddNoteRequest = z.infer<typeof AddNoteRequestSchema>;

/**
 * Request to search for CRM contacts.
 */
export const SearchContactsRequestSchema = z.object({
  query: z.string().optional().describe('Search query string'),
  fields: z.array(z.string()).optional().describe('Fields to return in results'),
  filters: z.record(z.any()).optional().describe('Additional filters to apply'),
  pagination: PaginationParamsSchema.optional().describe('Pagination parameters'),
});
export type SearchContactsRequest = z.infer<typeof SearchContactsRequestSchema>;

/**
 * A single contact search result with relevance score.
 */
export const ContactSearchMatchSchema = z.object({
  id: z.string().describe('Contact ID'),
  email: z.string().email().describe('Contact email'),
  first_name: z.string().optional().describe('First name'),
  last_name: z.string().optional().describe('Last name'),
  company: z.string().optional().describe('Company'),
  title: z.string().optional().describe('Job title'),
  phone: z.string().optional().describe('Phone'),
  score: z.number().min(0).max(1).describe('Relevance score (0-1)'),
  url: z.string().url().optional().describe('URL to view in provider system'),
});
export type ContactSearchMatch = z.infer<typeof ContactSearchMatchSchema>;

/**
 * Response from contact search operation.
 */
export const SearchContactsResponseSchema = z.object({
  matches: z.array(ContactSearchMatchSchema).describe('Search results'),
  pagination: PaginationResponseSchema.optional().describe('Pagination metadata'),
});
export type SearchContactsResponse = z.infer<typeof SearchContactsResponseSchema>;

File: packages/schema/src/python/crm.py
"""
CRM schema models for Transform Army AI Adapter Service.

This module defines Pydantic models for CRM operations including contacts,
companies, deals, and notes across different CRM providers (HubSpot, Salesforce, etc.).
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field, EmailStr, HttpUrl, ConfigDict, field_validator

from .base import ToolInput, PaginationParams, PaginationResponse


class Contact(BaseModel):
    """
    CRM contact model representing a person in the CRM system.
    
    This model abstracts contact data across different CRM providers.
    """
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": "cont_12345",
                "email": "john.doe@example.com",
                "first_name": "John",
                "last_name": "Doe",
                "company": "Acme Corp",
                "phone": "+1-555-0123",
                "title": "VP of Sales",
                "url": "https://app.hubspot.com/contacts/12345",
                "created_at": "2025-10-31T01:17:00Z",
                "updated_at": "2025-10-31T01:17:00Z",
                "custom_fields": {
                    "lead_source": "website",
                    "lead_score": 85
                }
            }
        }
    )
    
    id: str = Field(description="Unique contact identifier")
    email: EmailStr = Field(description="Contact email address")
    first_name: Optional[str] = Field(
        default=None,
        description="Contact first name"
    )
    last_name: Optional[str] = Field(
        default=None,
        description="Contact last name"
    )
    company: Optional[str] = Field(
        default=None,
        description="Company name"
    )
    phone: Optional[str] = Field(
        default=None,
        description="Phone number"
    )
    title: Optional[str] = Field(
        default=None,
        description="Job title"
    )
    url: Optional[HttpUrl] = Field(
        default=None,
        description="URL to view contact in provider system"
    )
    created_at: Optional[datetime] = Field(
        default=None,
        description="Contact creation timestamp"
    )
    updated_at: Optional[datetime] = Field(
        default=None,
        description="Last update timestamp"
    )
    custom_fields: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Provider-specific custom fields"
    )
    owner_id: Optional[str] = Field(
        default=None,
        description="ID of the contact owner/assignee"
    )
    tags: Optional[List[str]] = Field(
        default=None,
        description="Tags associated with the contact"
    )


class Company(BaseModel):
    """
    CRM company/account model.
    
    Represents an organization in the CRM system.
    """
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": "comp_456",
                "name": "Acme Corp",
                "domain": "acme.com",
                "industry": "Technology",
                "employees": 500,
                "annual_revenue": 10000000,
                "url": "https://app.hubspot.com/companies/456"
            }
        }
    )
    
    id: str = Field(description="Unique company identifier")
    name: str = Field(description="Company name")
    domain: Optional[str] = Field(
        default=None,
        description="Company website domain"
    )
    industry: Optional[str] = Field(
        default=None,
        description="Industry/sector"
    )
    employees: Optional[int] = Field(
        default=None,
        ge=0,
        description="Number of employees"
    )
    annual_revenue: Optional[float] = Field(
        default=None,
        ge=0,
        description="Annual revenue"
    )
    phone: Optional[str] = Field(
        default=None,
        description="Company phone number"
    )
    address: Optional[str] = Field(
        default=None,
        description="Company address"
    )
    city: Optional[str] = Field(
        default=None,
        description="City"
    )
    state: Optional[str] = Field(
        default=None,
        description="State/province"
    )
    country: Optional[str] = Field(
        default=None,
        description="Country"
    )
    postal_code: Optional[str] = Field(
        default=None,
        description="Postal/ZIP code"
    )
    url: Optional[HttpUrl] = Field(
        default=None,
        description="URL to view company in provider system"
    )
    created_at: Optional[datetime] = Field(
        default=None,
        description="Company creation timestamp"
    )
    updated_at: Optional[datetime] = Field(
        default=None,
        description="Last update timestamp"
    )
    custom_fields: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Provider-specific custom fields"
    )


class Deal(BaseModel):
    """
    CRM deal/opportunity model.
    
    Represents a sales opportunity or deal in the CRM pipeline.
    """
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": "deal_789",
                "name": "Acme Corp - Enterprise Plan",
                "amount": 50000,
                "currency": "USD",
                "stage": "qualification",
                "probability": 0.25,
                "close_date": "2025-12-31",
                "contact_ids": ["cont_12345"],
                "company_id": "comp_456"
            }
        }
    )
    
    id: str = Field(description="Unique deal identifier")
    name: str = Field(description="Deal name")
    amount: Optional[float] = Field(
        default=None,
        ge=0,
        description="Deal amount/value"
    )
    currency: str = Field(
        default="USD",
        description="Currency code (ISO 4217)"
    )
    stage: str = Field(
        description="Current deal stage/status"
    )
    probability: Optional[float] = Field(
        default=None,
        ge=0,
        le=1,
        description="Win probability (0-1)"
    )
    close_date: Optional[str] = Field(
        default=None,
        description="Expected close date (YYYY-MM-DD)"
    )
    contact_ids: Optional[List[str]] = Field(
        default=None,
        description="Associated contact IDs"
    )
    company_id: Optional[str] = Field(
        default=None,
        description="Associated company ID"
    )
    owner_id: Optional[str] = Field(
        default=None,
        description="Deal owner/assignee ID"
    )
    url: Optional[HttpUrl] = Field(
        default=None,
        description="URL to view deal in provider system"
    )
    created_at: Optional[datetime] = Field(
        default=None,
        description="Deal creation timestamp"
    )
    updated_at: Optional[datetime] = Field(
        default=None,
        description="Last update timestamp"
    )
    custom_fields: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Provider-specific custom fields"
    )
    
    @field_validator('currency')
    @classmethod
    def validate_currency(cls, v: str) -> str:
        """Validate currency code format (3 uppercase letters)."""
        if len(v) != 3 or not v.isupper():
            raise ValueError("Currency must be a 3-letter ISO 4217 code (e.g., USD, EUR)")
        return v


class Note(BaseModel):
    """
    CRM note/activity model.
    
    Represents a note or activity associated with a contact, company, or deal.
    """
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": "note_678",
                "content": "Initial qualification call completed.",
                "type": "call_note",
                "contact_id": "cont_12345",
                "created_at": "2025-10-31T01:15:00Z"
            }
        }
    )
    
    id: str = Field(description="Unique note identifier")
    content: str = Field(description="Note content/body")
    type: Optional[str] = Field(
        default=None,
        description="Note type (e.g., 'call_note', 'email', 'meeting')"
    )
    contact_id: Optional[str] = Field(
        default=None,
        description="Associated contact ID"
    )
    company_id: Optional[str] = Field(
        default=None,
        description="Associated company ID"
    )
    deal_id: Optional[str] = Field(
        default=None,
        description="Associated deal ID"
    )
    author_id: Optional[str] = Field(
        default=None,
        description="Note author/creator ID"
    )
    timestamp: Optional[datetime] = Field(
        default=None,
        description="Note timestamp (when the activity occurred)"
    )
    created_at: Optional[datetime] = Field(
        default=None,
        description="Note creation timestamp in system"
    )


class CreateContactRequest(ToolInput):
    """
    Request to create a new CRM contact.
    
    Includes deduplication options to handle existing contacts.
    """
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "idempotency_key": "idm_unique123",
                "correlation_id": "cor_req123",
                "contact": {
                    "email": "john.doe@example.com",
                    "first_name": "John",
                    "last_name": "Doe",
                    "company": "Acme Corp",
                    "phone": "+1-555-0123",
                    "title": "VP of Sales"
                },
                "options": {
                    "dedupe_by": ["email"],
                    "update_if_exists": True
                }
            }
        }
    )
    
    class ContactData(BaseModel):
        """Contact data for creation."""
        email: EmailStr = Field(description="Contact email (required)")
        first_name: Optional[str] = Field(default=None, description="First name")
        last_name: Optional[str] = Field(default=None, description="Last name")
        company: Optional[str] = Field(default=None, description="Company name")
        phone: Optional[str] = Field(default=None, description="Phone number")
        title: Optional[str] = Field(default=None, description="Job title")
        owner_id: Optional[str] = Field(default=None, description="Owner ID")
        tags: Optional[List[str]] = Field(default=None, description="Tags")
        custom_fields: Optional[Dict[str, Any]] = Field(
            default=None,
            description="Custom fields"
        )
    
    class ContactOptions(BaseModel):
        """Options for contact creation."""
        dedupe_by: Optional[List[str]] = Field(
            default=None,
            description="Fields to check for duplicates (e.g., ['email'])"
        )
        update_if_exists: bool = Field(
            default=False,
            description="Update contact if duplicate found"
        )
    
    contact: ContactData = Field(description="Contact data to create")
    options: Optional[ContactOptions] = Field(
        default=None,
        description="Creation options"
    )


class UpdateContactRequest(ToolInput):
    """Request to update an existing CRM contact."""
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "idempotency_key": "idm_update456",
                "correlation_id": "cor_req124",
                "updates": {
                    "title": "Senior VP of Sales",
                    "phone": "+1-555-0199"
                }
            }
        }
    )
    
    updates: Dict[str, Any] = Field(
        description="Fields to update (partial update)"
    )


class CreateDealRequest(ToolInput):
    """Request to create a new CRM deal/opportunity."""
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "idempotency_key": "idm_deal789",
                "correlation_id": "cor_req126",
                "deal": {
                    "name": "Acme Corp - Enterprise Plan",
                    "amount": 50000,
                    "currency": "USD",
                    "stage": "qualification",
                    "close_date": "2025-12-31",
                    "contact_ids": ["cont_12345"],
                    "company_id": "comp_456"
                }
            }
        }
    )
    
    class DealData(BaseModel):
        """Deal data for creation."""
        name: str = Field(description="Deal name")
        amount: Optional[float] = Field(default=None, ge=0, description="Deal amount")
        currency: str = Field(default="USD", description="Currency code")
        stage: str = Field(description="Deal stage")
        probability: Optional[float] = Field(
            default=None,
            ge=0,
            le=1,
            description="Win probability"
        )
        close_date: Optional[str] = Field(
            default=None,
            description="Expected close date (YYYY-MM-DD)"
        )
        contact_ids: Optional[List[str]] = Field(
            default=None,
            description="Associated contact IDs"
        )
        company_id: Optional[str] = Field(
            default=None,
            description="Associated company ID"
        )
        owner_id: Optional[str] = Field(default=None, description="Deal owner ID")
        custom_fields: Optional[Dict[str, Any]] = Field(
            default=None,
            description="Custom fields"
        )
    
    deal: DealData = Field(description="Deal data to create")


class UpdateDealRequest(ToolInput):
    """Request to update an existing CRM deal."""
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "idempotency_key": "idm_deal_update890",
                "correlation_id": "cor_req127",
                "updates": {
                    "stage": "negotiation",
                    "probability": 0.75,
                    "amount": 55000
                }
            }
        }
    )
    
    updates: Dict[str, Any] = Field(
        description="Fields to update (partial update)"
    )


class AddNoteRequest(ToolInput):
    """Request to add a note to a CRM entity."""
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "idempotency_key": "idm_note456",
                "correlation_id": "cor_req124",
                "note": {
                    "content": "Initial qualification call completed.",
                    "type": "call_note",
                    "timestamp": "2025-10-31T01:15:00Z"
                }
            }
        }
    )
    
    class NoteData(BaseModel):
        """Note data for creation."""
        content: str = Field(description="Note content")
        type: Optional[str] = Field(default=None, description="Note type")
        timestamp: Optional[datetime] = Field(
            default=None,
            description="When the activity occurred"
        )
    
    note: NoteData = Field(description="Note data to create")


class SearchContactsRequest(BaseModel):
    """Request to search for CRM contacts."""
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "query": "john.doe@example.com",
                "fields": ["email", "first_name", "last_name", "company"],
                "filters": {
                    "company": "Acme Corp",
                    "tags": ["lead", "qualified"]
                },
                "pagination": {
                    "page": 1,
                    "page_size": 10
                }
            }
        }
    )
    
    query: Optional[str] = Field(
        default=None,
        description="Search query string"
    )
    fields: Optional[List[str]] = Field(
        default=None,
        description="Fields to return in results"
    )
    filters: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Additional filters to apply"
    )
    pagination: Optional[PaginationParams] = Field(
        default=None,
        description="Pagination parameters"
    )


class ContactSearchMatch(BaseModel):
    """A single contact search result with relevance score."""
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": "cont_12345",
                "email": "john.doe@example.com",
                "first_name": "John",
                "last_name": "Doe",
                "company": "Acme Corp",
                "score": 0.98
            }
        }
    )
    
    id: str = Field(description="Contact ID")
    email: EmailStr = Field(description="Contact email")
    first_name: Optional[str] = Field(default=None, description="First name")
    last_name: Optional[str] = Field(default=None, description="Last name")
    company: Optional[str] = Field(default=None, description="Company")
    title: Optional[str] = Field(default=None, description="Job title")
    phone: Optional[str] = Field(default=None, description="Phone")
    score: float = Field(
        ge=0,
        le=1,
        description="Relevance score (0-1)"
    )
    url: Optional[HttpUrl] = Field(
        default=None,
        description="URL to view in provider system"
    )


class SearchContactsResponse(BaseModel):
    """Response from contact search operation."""
    
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "matches": [
                    {
                        "id": "cont_12345",
                        "email": "john.doe@example.com",
                        "first_name": "John",
                        "last_name": "Doe",
                        "company": "Acme Corp",
                        "score": 0.98
                    }
                ],
                "pagination": {
                    "page": 1,
                    "page_size": 10,
                    "total_pages": 1,
                    "total_items": 1,
                    "has_next": False,
                    "has_previous": False
                }
            }
        }
    )
    
    matches: List[ContactSearchMatch] = Field(description="Search results")
    pagination: Optional[PaginationResponse] = Field(
        default=None,
        description="Pagination metadata"
    )